File: extensions/resize.py
import pywinctl
import pywintypes
import keyboard
import time
import threading
import config
from animations import animate_to
from wutilerror import check_types

class Extension:
    def __init__(self):
        self.name = "resize"
        self.desc = "Resize windows smoothly by side, axis, or all sides equally."
        self.args = ["side", "amount"]
        self.short = "res"

    def main(self, window, side=None, amount=None):
        check_types(
            side=(side, str)
        )
        try:
            if window is None:
                title = input("Window title: ")
                matches = [w for w in pywinctl.getAllWindows() if title.lower() in w.title.lower()]
                if not matches:
                    print("No matching window found.")
                    return
                window = matches[0]

            if not side:
                print("Missing required argument: side")
                return window

            side = side.lower()
            dirs = {
                "left": ["left", "l"],
                "right": ["right", "r"],
                "top": ["top", "up", "u"],
                "bottom": ["bottom", "down", "d"],
                "all": ["all", "equal", "eq"],
                "horizontal": ["horizontal", "hor", "x"],
                "vertical": ["vertical", "ver", "vert", "y"]
            }

            direction = None
            for key, aliases in dirs.items():
                if side in aliases:
                    direction = key
                    break
            if direction is None:
                print(f"Unknown side '{side}'")
                return window

            if amount is not None:
                self._resize_once(window, direction, int(amount))
            else:
                print(f"Interactive resize mode ({direction}). Use arrow keys, ESC to exit.")
                self._interactive_resize(window, direction)

        except pywintypes.error:
            print("Window is no longer valid (likely closed).")
        except Exception as e:
            print(f"An error occurred: {e}")

        return window

    def _resize_once(self, window, direction, delta):
        # Direct one-shot resize (simple version)
        l, t, w, h = window.left, window.top, window.width, window.height

        if direction == "left":
            l -= delta
            w += delta
        elif direction == "right":
            w += delta
        elif direction == "top":
            t -= delta
            h += delta
        elif direction == "bottom":
            h += delta
        elif direction == "horizontal":
            l -= delta // 2
            w += delta
        elif direction == "vertical":
            t -= delta // 2
            h += delta
        elif direction == "all":
            l -= delta // 2
            t -= delta // 2
            w += delta
            h += delta

        animate_to(window, l, t,
                   duration=getattr(config, "RESIZE_ANIM_DURATION", 0.15),
                   steps=getattr(config, "RESIZE_ANIM_STEPS", 15))
        window.resizeTo(w, h)

    def _interactive_resize(self, window, direction):
        resize_amount = getattr(config, "DEFAULT_RESIZE_STRENGTH", 10)
        duration = getattr(config, "RESIZE_ANIM_DURATION", 0.1)
        steps = getattr(config, "RESIZE_ANIM_STEPS", 10)

        target = {
            "left": window.left,
            "top": window.top,
            "width": window.width,
            "height": window.height,
            "stop": False
        }

        def anim_loop():
            while not target["stop"]:
                animate_to(window, target["left"], target["top"], duration=duration, steps=steps)
                window.resizeTo(int(target["width"]), int(target["height"]))
                time.sleep(0.01)

        def handle_input():
            while True:
                if keyboard.is_pressed("esc"):
                    target["stop"] = True
                    print("Exiting resize mode.")
                    break

                l, t, w, h = target["left"], target["top"], target["width"], target["height"]

                # --- per-side behaviour ---
                if direction == "left":
                    if keyboard.is_pressed("left"):  # grow left
                        l -= resize_amount
                        w += resize_amount
                    if keyboard.is_pressed("right"):  # shrink left
                        l += resize_amount
                        w -= resize_amount

                elif direction == "right":
                    if keyboard.is_pressed("right"):  # grow right
                        w += resize_amount
                    if keyboard.is_pressed("left"):  # shrink right
                        w -= resize_amount

                elif direction in ("top",):
                    if keyboard.is_pressed("up"):  # grow upward
                        t -= resize_amount
                        h += resize_amount
                    if keyboard.is_pressed("down"):  # shrink upward
                        t += resize_amount
                        h -= resize_amount

                elif direction in ("bottom",):
                    if keyboard.is_pressed("down"):  # grow downward
                        h += resize_amount
                    if keyboard.is_pressed("up"):  # shrink downward
                        h -= resize_amount

                elif direction == "horizontal":
                    if keyboard.is_pressed("right"):  # grow both sides
                        l -= resize_amount // 2
                        w += resize_amount
                    if keyboard.is_pressed("left"):  # shrink both sides
                        l += resize_amount // 2
                        w -= resize_amount

                elif direction == "vertical":
                    if keyboard.is_pressed("down"):  # grow both vertically
                        t -= resize_amount // 2
                        h += resize_amount
                    if keyboard.is_pressed("up"):  # shrink both vertically
                        t += resize_amount // 2
                        h -= resize_amount

                elif direction == "all":
                    if keyboard.is_pressed("down") or keyboard.is_pressed("left"): #shrink all
                        l += resize_amount // 2
                        t += resize_amount // 2
                        w -= resize_amount
                        h -= resize_amount
                    if keyboard.is_pressed("up") or keyboard.is_pressed("right"): #grow all
                        l -= resize_amount // 2
                        t -= resize_amount // 2
                        w += resize_amount
                        h += resize_amount

                target.update(left=l, top=t, width=w, height=h)
                time.sleep(0.05)

        anim_thread = threading.Thread(target=anim_loop, daemon=True)
        anim_thread.start()
        handle_input()
        anim_thread.join()

File: extensions/resize.go
package main

//wutil:name resize
//wutil:short res
//wutil:desc Resize windows smoothly by side, axis, or all sides equally.
//wutil:args side,amount
//wutil:requires_window true

import (
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"

	"windowutil/pkg/wutilext"
	"windowutil/pkg/wutilwin"
)

func main() {
	side := ""
	if len(os.Args) > 1 {
		side = strings.ToLower(os.Args[1])
	}
	amount, hasAmount := optionalInt(2)
	wutilext.Mark("resize start", "resize")
	wutilext.LogDetail(fmt.Sprintf("requested resize side=%s amount=%v", side, maybeInt(amount, hasAmount)), "resize")

	win, err := wutilwin.RequireSelected("resize")
	if err != nil {
		fmt.Println(err)
		return
	}
	direction := resolveDirection(side)
	if direction == "" {
		if side == "" {
			fmt.Println("Missing required argument: side")
		} else {
			fmt.Printf("Unknown side '%s'\n", side)
		}
		wutilext.Log(fmt.Sprintf("resize direction could not be resolved from side=%s", side), "resize")
		return
	}
	if hasAmount {
		wutilext.Log(fmt.Sprintf("running one-off resize direction=%s delta=%d", direction, amount), "resize")
		resizeOnce(*win, direction, amount)
	} else {
		fmt.Printf("Interactive resize mode (%s). Use arrow keys, ESC to exit.\n", direction)
		wutilext.Log(fmt.Sprintf("starting interactive resize direction=%s", direction), "resize")
		interactiveResize(*win, direction)
	}
	wutilext.Mark("resize complete", "resize")
}

func resolveDirection(side string) string {
	aliases := map[string]string{
		"left": "left", "l": "left",
		"right": "right", "r": "right",
		"top": "top", "up": "top", "u": "top",
		"bottom": "bottom", "down": "bottom", "d": "bottom",
		"all": "all", "equal": "all", "eq": "all",
		"horizontal": "horizontal", "hor": "horizontal", "x": "horizontal",
		"vertical": "vertical", "ver": "vertical", "vert": "vertical", "y": "vertical",
	}
	return aliases[side]
}

func resizeOnce(win wutilwin.Window, direction string, delta int) {
	bounds, err := win.Bounds()
	if err != nil {
		fmt.Println("Window is no longer valid (likely closed).")
		wutilext.Log("window handle became invalid during resize", "resize")
		return
	}
	left, top, width, height := resized(bounds, direction, delta)
	animateMoveResize(win, bounds, left, top, width, height, 150*time.Millisecond, 15)
	wutilext.Log(fmt.Sprintf("_resize_once applied left=%d top=%d width=%d height=%d", left, top, width, height), "resize")
}

func interactiveResize(win wutilwin.Window, direction string) {
	for {
		if wutilwin.EscapePressed() {
			fmt.Println("Exiting resize mode.")
			wutilext.Log("interactive resize stopped by esc", "resize")
			return
		}
		delta := 0
		if wutilwin.RightPressed() || wutilwin.UpPressed() {
			delta = 10
		}
		if wutilwin.LeftPressed() || wutilwin.DownPressed() {
			delta = -10
		}
		if delta != 0 {
			bounds, err := win.Bounds()
			if err == nil {
				left, top, width, height := resized(bounds, direction, delta)
				_ = win.MoveResize(left, top, width, height)
				wutilext.LogDetail(fmt.Sprintf("interactive resize target left=%d top=%d width=%d height=%d", left, top, width, height), "resize")
			}
			time.Sleep(50 * time.Millisecond)
		}
		time.Sleep(10 * time.Millisecond)
	}
}

func resized(bounds wutilwin.Bounds, direction string, delta int) (int, int, int, int) {
	left, top, width, height := bounds.Left, bounds.Top, bounds.Width, bounds.Height
	switch direction {
	case "left":
		left -= delta
		width += delta
	case "right":
		width += delta
	case "top":
		top -= delta
		height += delta
	case "bottom":
		height += delta
	case "horizontal":
		left -= delta / 2
		width += delta
	case "vertical":
		top -= delta / 2
		height += delta
	case "all":
		left -= delta / 2
		top -= delta / 2
		width += delta
		height += delta
	}
	if width < 50 {
		width = 50
	}
	if height < 50 {
		height = 50
	}
	return left, top, width, height
}

func animateMoveResize(win wutilwin.Window, from wutilwin.Bounds, left int, top int, width int, height int, duration time.Duration, steps int) {
	if steps <= 0 {
		_ = win.MoveResize(left, top, width, height)
		return
	}
	sleep := duration / time.Duration(steps)
	for i := 1; i <= steps; i++ {
		t := float64(i) / float64(steps)
		x := from.Left + int(float64(left-from.Left)*t)
		y := from.Top + int(float64(top-from.Top)*t)
		w := from.Width + int(float64(width-from.Width)*t)
		h := from.Height + int(float64(height-from.Height)*t)
		_ = win.MoveResize(x, y, w, h)
		time.Sleep(sleep)
	}
}

func optionalInt(index int) (int, bool) {
	if len(os.Args) <= index {
		return 0, false
	}
	value, err := strconv.Atoi(os.Args[index])
	if err != nil {
		return 0, false
	}
	return value, true
}

func maybeInt(value int, ok bool) any {
	if !ok {
		return nil
	}
	return value
}
